content/docs/userspace/gall-guide/11-fail.md

```diff
@@ -0,0 +1,139 @@
+`on-fail` arm for handling. This arm is very seldom used, almost all agents
+leave it for `default-agent` to handle, which just prints the error message to
+the terminal. While you're unlikely to use this arm, we'll briefly go over its
+behavior for completeness.
+
+`on-fail` takes a `term` error message and a `tang`, typically containing a
+stack trace, and often with additional messages about the error. If it weren't
+delegated to `on-fail:def`, it would begin with:
+
+```hoon
+++  on-fail
+  |=  [=term =tang]
+  ^-  (quip card _this)
+  ....
+```
+
+Gall calls `on-fail` in four cases:
+
+- When there's a crash in the `on-arvo` arm.
+- When there's a crash in the `on-agent` arm.
+- When there's a crash in the `on-leave` arm.
+- When an agent produces a `%watch` card but the `wire`, ship, agent and `path`
+  specified are the same as an existing subscription.
+
+For an `on-arvo` failure, the `term` will always be `%arvo-response`, and the
+`tang` will contain a stack trace.
+
+For `on-agent`, the `term` will be the head of the `sign` (`%poke-ack`, `%fact`,
+etc). The `tang` will contain a stack trace and a message of "closing
+subscription".
+
+For an `on-leave` failure, the `term` will always be `%leave`, and the `tang`
+will contain a stack trace.
+
+For a `%watch` failure, the `term` will be `%watch-not-unique`. The `tang` will
+include a message of "subscribe wire not unique", as well as the agent name, the
+`wire`, the target ship and the target agent.
+
+How you might handle these cases (if you wanted to manually handle them) depends
+on the logic of your particular agent.
+
+## Helper core
+
+Back in the lesson on lustar virtual arms, we briefly mentioned a common pattern
+is to define an alias of `hc` for a "helper core" like:
+
+```hoon
++*  this  .
+    def   ~(. (default-agent this %.n) bowl)
+    hc    ~(. +> bowl)
+```
+
+A helper core is a separate core composed into the subject of the agent core,
+containing useful functions for use by the agent arms. Such a helper core would
+typically contain functions that would only ever be used internally by the
+agent - more general functions would usually be included in a separate `/lib`
+library and imported with a [faslus](/docs/arvo/ford/ford#ford-runes) (`/+`)
+rune.
+
+The conventional pattern is to have the helper core _below_ the agent core, so the structure of the agent file is like:
+
+```
+[imports]
+[state types core]
+[agent core]
+[helper core]
+```
+
+Recall that the build system will implicitly compose any discrete expressions.
+If we simply added the helper core below the agent core, the agent core would be
+composed into the subject of the helper core, which is the opposite of what we
+want. Instead, we must inversely compose the two cores with a
+[tisgal](/docs/hoon/reference/rune/tis#-tisgal) (`=<`) rune. We add the tisgal
+rune directly above the agent core like:
+
+```hoon
+.....
+=<
+|_  =bowl:gall
++*  this      .
+    def   ~(. (default-agent this %|) bowl)
```

Comment:
```suggestion
    def   ~(. (default-agent this %.n) bowl)
```
keeping things consistent with the rest of the guide (also I prefer this over `%|`)